homeassistant/components/rainmachine/__init__.py

```diff
@@ -128,6 +130,39 @@
 }, extra=vol.ALLOW_EXTRA)
 
 
+async def _check_service_permissions(hass, service_call):
+    """Ensure the user of a service call has proper permissions."""
+    if service_call.data.get('entity_id'):
+        rainmachine_entities = service_call.data['entity_id']
+    else:
+        # If the service call doesn't explicitly provide entity IDs, use _all_
+        # RainMachine entities:
+        en_reg = await hass.helpers.entity_registry.async_get_registry()
+        rainmachine_entities = [
+            entity.entity_id for entity in en_reg.entities
+            if hass.config_entries.async_get_entry(
+                entity.config_entry_id).domain == DOMAIN
+        ]
+
+    for entity_id in rainmachine_entities:
+        if service_call.context.user_id:
+            user = await hass.auth.async_get_user(service_call.context.user_id)
+
+            if user is None:
```

Comment:
No need to do this for each entity. Also move it up before even fetching any entities.